lib/galaxy/managers/library_datasets.py

```diff
@@ -0,0 +1,234 @@
+                if datatype is None:
+                    raise RequestParameterInvalidException('This Galaxy does not recognize the datatype of: %s' % (val))
+                validated_payload[key] = val
+            if key in ('genome_build'):
+                if len(val) < MINIMUM_STRING_LENGTH:
+                    raise RequestParameterInvalidException('%s must have at least length of %s' % (key, MINIMUM_STRING_LENGTH))
+                val = validation.validate_and_sanitize_basestring(key, val)
+                validated_payload[key] = val
+        return validated_payload
+
+    def secure(self, trans, ld, check_accessible=True, check_ownership=False):
+        """
+        Check if library dataset is accessible to current user or the user is an admin.
+
+        :param  ld:         library dataset
+        :type   ld:         galaxy.model.LibraryDataset
+        :param  check_accessible:        flag whether to check that user can access library dataset
+        :type   check_accessible:        bool
+
+        :returns:   the original library dataset
+        :rtype:     galaxy.model.LibraryDataset
+        """
+        if trans.user_is_admin():
+            # all operations are available to an admin
+            return ld
+        if check_accessible:
+            ld = self.check_accessible(trans, ld)
+        return ld
+
+    def check_accessible(self, trans, ld):
+        """
+        Check whether the current user has permissions to access library dataset.
+
+        :param  ld: library dataset
+        :type   ld: galaxy.model.LibraryDataset
+
+        :returns:   the original library dataset
+        :rtype:     galaxy.model.LibraryDataset
+
+        :raises:    ObjectNotFound
+        """
+        if not trans.app.security_agent.can_access_library_item(trans.get_current_user_roles(), ld, trans.user):
+            raise ObjectNotFound('Library dataset with the id provided was not found.')
+        elif ld.deleted:
+            raise ObjectNotFound('Library dataset with the id provided is deleted.')
+        else:
+            return ld
+
+    def check_modifiable(self, trans, ld):
+        """
+        Check whether the current user has permissions to modify library dataset.
+
+        :param  ld: library dataset
+        :type   ld: galaxy.model.LibraryDataset
+
+        :returns:   the original library dataset
+        :rtype:     galaxy.model.LibraryDataset
+
+        :raises:    ObjectNotFound
+        """
+        if trans.user_is_admin():
+            # all operations are available to an admin
+            return ld
+        if not trans.app.security_agent.can_modify_library_item(trans.get_current_user_roles(), ld):
+            raise InsufficientPermissionsException('You do not have proper permission to modify this library dataset.')
+        elif ld.deleted:
+            raise ObjectNotFound('Library dataset with the id provided is deleted.')
```

Comment:
I think the check on `deleted` should be the first, admins are not allowed to modify a deleted library dataset anyway by the check on line 62. So I think this should be:
```python
    if ld.deleted:
        raise ...
    if trans.user_is_admin():
        return ld
    if not trans.app.security_agent...
        raise ...
```